content/200-concepts/100-components/02-prisma-client/052-extend-client.mdx

```diff
@@ -0,0 +1,401 @@
+  const userWithoutPassword = exclude(user, 'password')
+}
+```
+
+### Exclude Password in Typescript
+
+Now let's make this `exclude` function type-safe. You first want to make sure that the keys you provide exist on the user and then you want to return a user that doesn't contain those keys.
+
+Here's what that would look like in Typescript:
+
+```tsx file=index.ts highlight=1-4;normal
+function exclude<User, Key extends keyof User>(
+  user: User,
+  ...keys: Key[]
+): Omit<User, Key> {
+  for (let key of keys) {
+    delete user[key]
+  }
+  return user
+}
+
+function main() {
+  const user = await prisma.user.findUnique({ where: 1 })
+  const userWithoutPassword = exclude(user, 'password')
+}
+```
+
+In this example, we've provided two generics `User` and `Key`. The `Key` generic is defined as the keys of a `User` (e.g. email, password, firstName, etc.). These generics flow through the logic, returning a User that omits the list of keys provided.
+
+## Organizing the Prisma Client **into Models**
+
+In the examples above, you've created helper functions that weave the type-safety deeper through your application. Creating helper functions works well when your application is small, but as your application grows, you may find an inconsistent use of these helper functions strewn across your codebase.
+
+One technique for consistently applying and organizing these helper functions is to group them in a class.
+
+### Wrapping the Model in a Class
+
+The most obvious way to wrap the Prisma Client is by wrapping it in a `Model` class:
+
+```tsx highlight=4,23,28-30;normal
+import { PrismaClient } from '@prisma/client'
+
+class Model {
+  constructor(private readonly model: PrismaClient['user']) {}
+
+  /**
+   * Signup the first user and create a new team of one. Return the User with
+   * a full name and without a password
+   */
+  async signup(input: Signup) {
+    assert(input, Signup)
+    const team = await prisma.team.create({
+      data: {
+        ...input.team,
+        users: {
+          create: input.user,
+        },
+      },
+      include: {
+        users: true,
+      },
+    })
+    return exclude(computeFullName(team.users[0]), 'password')
+  }
+}
+
+async function main() {
+  const prisma = new PrismaClient()
+  const model = new Model(prisma.user)
+  const user = await model.signup(/* { ... } */)
+  console.log(user)
+}
+```
+
+Here we're only exposing a `signup` method from the Prisma Client. With this approach, you'll no longer be able to call Prisma Client methods like findMany or upsert. This approach works well when you have a large application and you want to intentionally limit what your models can do.
+
+### Extending the Prisma Client
+
+But what if you don't want to hide existing functionality? In this case, you can use `Object.assign` to extend the Prisma Client without limiting it's functionality:
+
+```tsx highlight=2,27-29;normal
+function Model(model: PrismaClient['user']) {
```

Comment:
```suggestion
import { PrismaClient } from '@prisma/client'
import { assert } from 'superstruct'

function Model(model: PrismaClient['user']) {
```
missing imports?

Reply:
Thanks, it gets a bit long to include everything, but I've added some ellipsis to explain the missing pieces.